keras_nlp/layers/cached_multi_head_attention.py

```diff
@@ -0,0 +1,116 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Cached MHA layer based on `keras.layers.MultiHeadAttention`."""
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.compiler.tf2xla.python.xla import dynamic_update_slice
+
+
+class CachedMultiHeadAttention(keras.layers.MultiHeadAttention):
+    """MutliHeadAttention layer with cache support.
+
+    In autoregressive decoding, it's a common practice to cache the key/value in
+    multi-head attention of previously seen tokens in order to make the
+    computation faster. With cached K and V, we can compute the attention output
+    of the last token without recomputing the forward pass for previously seen
+    tokens. This caching method is only useful during decoding, and should not
+    be used during training.
+
+    Call arguments:
+        query: Query `Tensor` of shape `(B, T, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`.
+        value: Value `Tensor` of shape `(B, S, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`.
+        key: Optional key `Tensor` of shape `(B, S, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`. If not given, will use `value` for both
+            `key` and `value`, which is the most common case.
+        attention_mask: a boolean mask of shape `(B, T, S)` if `cache=None`,
+            otherwise `(B, 1, S)`. `attention_mask` prevents
+            attention to certain positions. The boolean mask specifies which
+            query elements can attend to which key elements, 1 indicates
+            attention and 0 indicates no attention. Broadcasting can happen for
+            the missing batch dimensions and the head dimension.
+        cache: a dense float Tensor. The cache of key/value of leading tokens.
+            `cache` is of shape [2, B, max_seq_len, num_heads, key_dims].
+        cache_index: a int or int Tensor, the index of the current token being
+            processed. If `cache_index=None` while `cache` is set, it means
+            it's the first pass to build the cache.
+
+    Returns:
+        attention_output: The result of the computation, of shape `(B, T, E)`,
+            where `T` is for target sequence shapes and `E` is the query input
+            last dimension if `output_shape` is `None`. Otherwise, the
+            multi-head outputs are projected to the shape specified by
+            `output_shape`.
+        cache: the updated cache.
+    """
+
+    def call(
+        self,
+        query,
+        value,
+        key=None,
+        attention_mask=None,
+        use_causal_mask=True,
```

Comment:
This feels weird to me, adding a cache should not change the default masking behavior.

Could we just remove this option and fixup the unit tests?

Reply:
Also, won't this mean in our current forward pass we will compute a causal mask twice and merge the same thing twice. That feels both confusing and inefficient!

Reply:
Surely the call method's signature should be identical to that of the parent except for cache related args.

Reply:
We should keep the default behavior one to one with the parent except for caching behavior.

I care less about supporting all the bells and whistles on the base class (e.g. supporting implicit masks with the cache might take some thought, and we have no direct use case yet).

Reply:
This is directly from the MHA: [link](https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/layers/attention/multi_head_attention.py#L550)

Should we stick to the parent class for now and reevaluating if `use_causal_mask` is required for MHA, if it is then I guess we should keep it for consistency.

Reply:
talked offline, we are flipping `use_causal_mask=False` for now, and put the rest into a followup PR.